mcs/class/System/System.IO.Compression/DeflateStream.cs

```diff
@@ -47,6 +48,8 @@ public class DeflateStream : Stream
 		bool leaveOpen;
 		bool disposed;
 		DeflateStreamNative native;
+		object disposeLock = new Object ();
+		CountdownEvent active_users = new CountdownEvent (1);
```

Comment:
It would be much better to use a SafeHandle to wrap `DeflateStreamNative.z_stream` so we make sure not to use it after it's been disposed. This would protect any races where you would use it after it's been disposed, as well as ensure that it is not destroyed while somebody is using it.